### ISSUE TITLE
Can't ssh using RBAC when VM doesn't have a public IP

### ISSUE THREAD

**Issue by rafagsiqueira** [Azure/azure-cli-extensions#3501]
- If the issue is to do with Azure CLI 2.0 in-particular, create an issue here at [Azure/azure-cli](https://github.com/Azure/azure-cli/issues)

### Extension name (the extension in question)
az ssh

### Description of issue (in as much detail as possible)
From the error it seems the az ssh command requires a public IP. Is that a requirement when using az ssh with RBAC?

``` The command failed with an unexpected error. Here is the traceback:
'NoneType' object has no attribute 'id'
Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/knack/cli.py", line 231, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 657, in execute
    raise ex
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 720, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 691, in _run_job
    result = cmd_copy(params)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 328, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/command_operation.py", line 121, in handler
    return op(**command_args)
  File "/home/rafael/.azure/cliextensions/ssh/azext_ssh/custom.py", line 22, in ssh_vm
    public_key_file, private_key_file, use_private_ip, ssh_utils.start_ssh_connection)
  File "/home/rafael/.azure/cliextensions/ssh/azext_ssh/custom.py", line 40, in _do_ssh_op
    ssh_ip = ssh_ip or ip_utils.get_ssh_ip(cmd, resource_group, vm_name, use_private_ip)
  File "/home/rafael/.azure/cliextensions/ssh/azext_ssh/ip_utils.py", line 27, in get_ssh_ip
    parsed_ip_id = tools.parse_resource_id(public_ip_ref.id)
AttributeError: 'NoneType' object has no attribute 'id'```
-----

**Comment by yonzhan** [comment]
vm ssh

**Comment by N6UDP** [comment]
We will get a better error and some documentation updates but this is largely by-design.  The ssh extension has no way to automatically reach / tunnel to VM instances with private IPs.  There are various strategies one can use to reach such instances (ssh forwarding, VPNs, tunneling, running on a host with access to the VM, etc all of which will work with the IP / config / cert ssh extension options.

**Comment by carlosporter** [comment]
Hello I just ran into this same problem, although there is one caveat on my case.

I have a jumpbox to access the virtual machines without public IP addresses and I can also install the az cli extension on it without problem.

Then I though that I cound export the configuration of the extension with "az ssh config --file configfile" and use that file later with the regular ssh client to reach the other virtual machine from the jumpbox.

Something like:

ssh -F ./configfile -l usuario@contoso.com  <private-ip-address>

So, the only problem, is that I don't see how the extension will create the user account for "usuario@contoso.com" if I'm not using the azure cli extension to access the server.

Any thoughts on that?

**Comment by N6UDP** [comment]
@carlosporter the user is created upon login by the extension running on the machine -- it's not created by the client -- the client only handles generating the cert and putting together an openssh command for connecting.

That said, we also can fall back to private ip by default.. we do have `--prefer-private-ip` for cases where you've brought your own connectivity today but it doesn't attempt a fallback.  I'm leaning toward producing an error message with information including the option as we don't do anything to help you get to a private ip today.

**Comment by carlosporter** [comment]
@N6UDP Thanks for the feedback,

I ended up finding the option "--prefer-private-ip" and I agree with you, this issue would be easier to trouble if there was a message alerting the user that the connection cannot be established because the VM targeted doesn't have public IP address.

You may also run into a problem if you still have a public IP address but access is restricted to an specific address via an NSG rule.

Given my experience with Azure so far, it's more likely to find virtual machines without a public IP address that are accessible only through a jumpbox or bastion server that is also located on Azure.

Also, it would be nice see a future version of the "az ssh" extension that includes support for agent forwarding to use a  Linux jumpbox and then reach any machines on a private subnet from it .

**Comment by N6UDP** [comment]
Fixed in #3698 and https://github.com/MicrosoftDocs/azure-docs/pull/78826

**Comment by maertendMSFT** [comment]
closing as it is called out as fixed.
